### PR TITLE
Make Message creation payload options more flexible

### PR DESCRIPTION
**Pull request by Gelbpunkt**

### CODE REVIEW

File: src/proto/types.rs [hunk 1]
```diff
@@ -262,6 +268,10 @@ enum PayloadStorage {
 ///
 /// Received messages are always validated prior to dealing with them, so all
 /// the type casting methods are either almost or fully zero cost.
+///
+/// When the client is on the end that has to mask frames, doing so will be
```
Comment (Erk-) [inline src/proto/types.rs:272]:
```suggestion
/// When the client is acting as a client and not as a server it has to mask frames, doing so will be
```

**Review by Erk- — APPROVED** [review]
Looks good, though maybe make the wording a bit more clear so users don't need to know what is the "end that has to mask frames"

**Comment by Gelbpunkt** [comment]
Still not sure what I think about this since `Into<Payload>` is a constraint on a type that isn't exposed. The docs explicitly state that the route via `Bytes` and `BytesMut` should be taken, so it might just be okay, but I am still not sure about it. Exposing `Payload` is not really an option because it implements things like a `TryInto<BytesMut>` impl that shouldn't be used. Thoughts @Erk- @vilgotf?

**Comment by Gelbpunkt** [comment]
> Still not sure what I think about this since `Into<Payload>` is a constraint on a type that isn't exposed. The docs explicitly state that the route via `Bytes` and `BytesMut` should be taken, so it might just be okay, but I am still not sure about it. Exposing `Payload` is not really an option because it implements things like a `TryInto<BytesMut>` impl that shouldn't be used. Thoughts @Erk- @vilgotf?

Additionally I've been thinking about using `transmute` hacks to replicate `BytesMut::from_vec` but I think that isn't really feasible in the long run. It would be nice for avoiding allocations, however.

**Comment by Erk-** [comment]
Maybe it is worth fixing this in a breaking release and then making a public version of Payload and then actually use a private type inside the library?

**Comment by Gelbpunkt** [comment]
> Maybe it is worth fixing this in a breaking release and then making a public version of Payload and then actually use a private type inside the library?

The more I think about it, the less I see a sense in hiding the trait impls of `Payload`. I would love to hear @vilgotf's opinion on this, but I've pushed what I think is a fairly reasonable approach to this PR.

File: src/proto/types.rs [hunk 2]
```diff
@@ -143,12 +143,23 @@ impl TryFrom<u16> for CloseCode {
...
+/// implementations.
+///
+/// Sending the payloads is zero-copy, except when sending a payload backed by
+/// [`Bytes`] as a client to a server. The use of [`BytesMut`] as the backing
+/// storage where cheaply possible is recommended to ensure zero-copy sending.
 ///
 /// [`Bytes`] offers no API for acquiring mutable access to *unique* slices.
-/// By defering calling `freeze` on [`BytesMut`], this type retains mutable
+/// By deferring calling `freeze` on [`BytesMut`], this type retains mutable
 /// access to the underlying bytes until access is shared.
```
Comment (vilgotf) [inline src/proto/types.rs:158]: This should not be public

**Review by vilgotf — COMMENTED** [review]
The `impl<'a> From<&'a Payload> for &'a Bytes` and `TryFrom<Payload> for BytesMut` implementations are very very much so internal code that I don't think should be exposed. These can, however, trivially be made into private methods on `Payload`.

I'm not super invested into this change, so I can't recommend an alternative approach

**Comment by Gelbpunkt** [comment]
> The `impl<'a> From<&'a Payload> for &'a Bytes` and `TryFrom<Payload> for BytesMut` implementations are very very much so internal code that I don't think should be exposed. These can, however, trivially be made into private methods on `Payload`.
> 
> I'm not super invested into this change, so I can't recommend an alternative approach

I've made the implementations into private methods. Could you take another look?

**Review by vilgotf — APPROVED** [review]

**Pull request merged**